independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java

```diff
@@ -735,13 +735,16 @@ static boolean hasQualifier(BeanDeployment beanDeployment, AnnotationInstance re
         return false;
     }
 
-    static void addImplicitQualifiers(Set<AnnotationInstance> qualifiers) {
-        if (qualifiers.isEmpty()
-                || (qualifiers.size() <= 2 && qualifiers.stream()
-                        .allMatch(a -> DotNames.NAMED.equals(a.name()) || DotNames.ANY.equals(a.name())))) {
+    static Set<AnnotationInstance> addImplicitQualifiers(Set<AnnotationInstance> qualifiers) {
+        if (qualifiers.isEmpty()) {
+            return BuiltinQualifier.DEFAULT_QUALIFIERS;
+        }
+        if ((qualifiers.size() <= 2 && qualifiers.stream()
+                .allMatch(a -> DotNames.NAMED.equals(a.name()) || DotNames.ANY.equals(a.name())))) {
             qualifiers.add(BuiltinQualifier.DEFAULT.getInstance());
         }
         qualifiers.add(BuiltinQualifier.ANY.getInstance());
+        return qualifiers;
```

Comment:
We could probably optimize things a bit more here and return the default set in more cases but I'm not sure it's worth the complexity.